Change run.stats file format to sqlite3

kren1 opened the pull request:
This pull request changes the file format used to store run.stats file to sqlite3.

Disadvantage:
 1. *run.stats no longer human readable*: I would consider the current format to be human-non-readable, but with this you can't open it in a text editor.

Advantages:
 1. *smaller run.stats file size*: Sqlite3 is a binary format which immediately reduces filesize. Furthermore it uses variable length encoding for integers, so it can use less than `numberOfRecords * numberOfFields * widthOfFields` bytes
1. *faster writes*: In most uses cases it should write faster than the current method (see below)
1. *faster klee-stats*: `klee-stats` gets very slow in my experience when dealing with `run.stats` with a lot of records, moving to sqlite3 should enable this to be much faster. I didn't benchmark  `klee-stats` in this PR, because that is a seprate issue
1. *SQL query lagnuage*: performing ad-hoc queries is much easier. You can just run `sqlite3 klee-last/run.stats` and type whatever SQL query you have in mind
1. *Well defined fail behaviour*:  I don't think it's currently well defined what stats are written if KLEE crashes. SQLIte enables us to control this (the trade-off is performance)


## Performance

I performed a small scale benchmark, similar to what KLEE does,, where I write 13 `int` records into a table. I can get 1M writes/s, but only about 1K/s commits on my machine. That means we can have very fast writes (that is writes on every instruction), but we can't commit on every instruction. 

So I came up with the following scheme:
* If we write on interval we commit for every write 
* If we write on instructions, we commit for every 1000 writes
* User can override the above settings with whatever they want

This gives sensible (fast) defaults, but still enables slow reliable per instruction writes.

I also ran a `od.bc` dumping a concrete file with ` -stats-write-after-instructions=1 -stats-write-interval=0`, both with and without this patch. For some reason me and @MartinNowack  couldn't get KLEE to execute the same number of instructions on this benchmark, but SQLite3 seems to do more work in the same time:

CSV:
```
KLEE: done: total instructions = 4656860
KLEE: done: completed paths = 1
KLEE: done: generated tests = 1

real    0m24.402s
user    0m18.408s
sys     0m5.980s
n$ ls -lh klee-last/run.stats
-rw-r--r-- 1 tk1713 srg 648M May 27 11:51 klee-last/run.stats
```

SQLite:
```
KLEE: done: total instructions = 5361184
KLEE: done: completed paths = 1
KLEE: done: generated tests = 1

real    0m24.787s
user    0m20.428s
sys     0m4.344s
$ ls -lh klee-last/run.stats
-rw-r--r-- 1 tk1713 srg 247M May 27 11:53 klee-last/run.stats
```

## Grafana 

This was inspired by @251 talk on Grafana. Unfortunately Grafana doesn't support SQLite as datasource (yet, maybe?), but I think this gets us closer to there. We just need to implement a [simple json datasource](https://github.com/grafana/simple-json-datasource), which is a webserver that queries the database a bit. I think this is much easier to do on top of SQLite than a CSV file. There is even a [blog post](https://nbsoftsolutions.com/blog/replacing-elasticsearch-with-rust-and-sqlite) about it.

This is now implemented to an extent. You can run `klee-stats -grafana path/to/a/single/klee-out-file`

The easiest way to run grafana with docker is 

```
docker run \
  -d \
  --net=host \
  --name=grafana \
  -e "GF_INSTALL_PLUGINS=grafana-clock-panel,grafana-simple-json-datasource" \
  grafana/grafana
```

And then create a SimpleJson datasource pointing to whatever klee-stats tells you (probably `http://localhost:5000`)

Screenshot of `od` running for a couple of minutes: 
![image](https://user-images.githubusercontent.com/310855/41111513-1077c17a-6a74-11e8-8e6b-3c0a9ae7f24b.png)


## klee-stats

I've only changed klee-stats in a minimal way to support what I think is the common use case. I don't want this to be bogged down because of `klee-stats`, so I think there are 3 options:

* Delete all the complicated code from `klee-stats` and support only the core.
* Conform to the current LazyList implementation, but this involved reading in the whole database and would give poor performance.
* Decide what nice to have features we want in `klee-stats` (without referring to the current implementation). I'm not sure what people use, but I would keep only the `--print-*` options and add some nice interface to SQL like `klee-stats -s 'SELECT * FROM stats WHERE ... ORDER BY ...'`

SQLite even support [adding custom functions](http://charlesleifer.com/blog/five-reasons-you-should-use-sqlite-in-2016/) to SQL, which could make querying in `klee-stats` even easier.

251 commented:
As I told you, I like the change. What I dislike (and what is orthogonal to this patch) is that we still have to change code in different places (corestats, writestatsheader, writestatsline, klee-stats,...) to add a simple statistic. Can't we use a more generic solution by e.g. adding an iterator to the StatisticManager?

For klee-stats: I'd default to `--print-all`, add `--print header-list` and just keep `--table-format`.

ccadar commented:
Thanks, @kren1.  Although for such changes it's better to discuss the design before implementing anything, to avoid wasted effort.   

I'm not sure I like a SQL solution for this -- what we discussed before was a structured text format such as YAML.  What do you think about that?

kren1 commented:
@251 Sure, but I don't think it fits in this PR. Also the problematic options are `--sort-by`, `--ascending`, `'--compare-by`,  `'--compare-at`, `--draw-line-chart'`.

@ccadar  We had a discusison about this on Friday after the meeting  @251 and @MartinNowack so I wanted to support my argument with some benchmarks, then I though I might as well do a PR out of it.

I don't think YAML solves the problem of storing and querying data, so I'm not sure what you mean, but here are some thought:

* Encoding rows in YAML format: That would make it a lot more verbose and perhaps more readable, but would make much larger files. Also it's schema-less, but we have a schema that doesn't change while KLEE is running. I don't see a reason we would want rows with different fields in the same run.stats file.

*  Maybe we could store YAML encoded rows in a NoSQL database, but KLEE doesn't generate nearly enough data to warant one. They are harder to setup and maintain (I don't want to install and setup one just to run KLEE). Querying is also more annoying usually.


SQLite is very close to the current implementation, it just gives formal structure to the data (which we alreadz have but don't enforce), it has a well known language to query it (so we don't have to reinvent the wheel with klee-stats) and is a sensible database (instead of the wild west with files)

251 commented:
@kren1 I agree. For the options: that's why I wouldn't keep them. ;)

@ccadar You mean as in #534? I think these are orthogonal and imho #534 is overly complex and could be simplified a lot with this patch. What's missing is a type info table for stats (duration, counter, memory) for different print modi (if needed).

251 commented:
Something that maps the underlying uint64_t to seconds, megabytes,... for fancy printing (again: if needed). You can do it by name (Instruction*Counter*, Solver*Time*) or by a mapping table. But this is more about a more generic solution where we don't know the number or names of statistics beforehand or where we want to integrate #534.

kren1 commented:
Ah I see, that's also orthogonal. In this solution the schema is created per run, so statistics and types can be changed at will basically.

251 commented on lib/Core/StatsTracker.cpp:
```diff
@@ -243,13 +257,20 @@ StatsTracker::StatsTracker(Executor &_executor, std::string _objectFilename,
+    //assert(statsFile && "unable to open statistics trace file");
+    if(sqlite3_open(executor.interpreterHandler->getOutputFilename("run.stats").c_str(), &statsFile)){
+       klee_error("Can't open database: %s", sqlite3_errmsg(statsFile));
+       sqlite3_close(statsFile);
```
`klee_error` calls `exit(1)` and `sqlite3_close()` is unreachable

ccadar commented:
@kren1 I took another look at this, and I'm getting more positive about it :)  I still need to try it though.  But can I quickly ask you whether it's backwards compatible with what we have now?  I.e., if you run `klee-stats` on several output directories, does it print the same thing?

kren1 commented:
@ccadar Great :D . Yes it is backward compatible in that manner.

The new features (grafana and CSV conversion) however don't do that. They just pick the first directory. I can probably get that working with grafana too if needed.

ccadar commented:
@kren1 If the main `klee-stats` tool will behave the same, that's good enough.  Let me know when you want a review of this (but please first group the changes into fewer nicer commits :))

kren1 commented:
@ccadar  I'm pretty happy with it now. I don't want to squash the `Remove <klee-stats option>` commits, if I need to revert them, but I can do it afterwards.

251 commented on CMakeLists.txt:
```diff
@@ -343,6 +343,12 @@ if (NOT LLVM_ENABLE_RTTI)
   endif()
 endif()
 
+find_package(SQLite3)
```
This should be a little further down and have a dedicated section with header similar to libcap/libutil/....

251 commented on lib/Core/CMakeLists.txt:
```diff
@@ -45,7 +45,7 @@ endif()
 
 klee_get_llvm_libs(LLVM_LIBS ${LLVM_COMPONENTS})
-target_link_libraries(kleeCore PUBLIC ${LLVM_LIBS})
+target_link_libraries(kleeCore PUBLIC ${LLVM_LIBS}  ${SQLITE3_LIBRARIES})
```
Wouldn't that add sqlite to kleaver?

kren1 replied:
Probably, but where would you put it to not do it? I don't want to move `StatsTracker` out of core

kren1 replied:
And leave kleaver without sql references? If kleeCore gets into kleaver it's not my fault.

251 replied:
Then keep it.

251 commented on lib/Core/StatsTracker.cpp:
```diff
@@ -76,11 +76,16 @@ namespace {
+  cl::opt<std::uint32_t> CommitEvery(
+      "stats-commit-after", cl::init(0),
+      cl::desc("Commit the statistics every N writes. Setting to 0 commits every write in interval mode (default) or every 1000 writes in write after N instructions"
+               "(default=0)"));
```
This needs rephrasing. I would expect
```
StatsWriteAfterInstructions CommitEvery
0 (off)                     0 (off)
0 (off)                     n (error)
n                           0 (write == commit)
n                           m (commit after intervals of n*m writes)

kren1 replied:
That's not how it works.
It's:

```
StatsWriteAfterInstructions        WriteInterval            CommitEvery
0 (off)                                                  n                           0 (commit every n seconds)
n                                                         m                          k (error)
0 (off)                                                  0                           n (off (why would I error?))
n                                                         0                           0 (commit every 1000 instructions)
n                                                         0                           m (commit after m writes)
```

If you don't specify CommitEvery and specify WriteAfterInstruction, your version would  not have sane defaults. If you are using WriteAfterInstructions and commit every write, you will most likely take a significant performance penality, which is not the default you want. I argued for that in the PR description.

251 replied:
I didn't specify the default values. Thinking about it, I would disable 0 for CommitEvery:
```
StatsWriteAfterInstructions WriteInterval CommitEvery
0                           0             _ (stats off, warn if CommitEvery set)
_                           _             0 (error, no commit)
0                           n             k (write every n seconds, commit every k writes)
m                           0             k (write every m instrs, commit every k writes)
m                           n             k (write every m instrs OR n seconds, commit every k writes)
```
Default values sth. like: 0/1s/30?

kren1 replied:
No, you need different default for write interval/instructions. Also I imagine most users don't want to deal with commitEvery.

251 replied:
I've only proposed a sane default for `CommitEvery` and kept the other ones. Proposing an instruction-based default is hard as this depends on the runtime behaviour of the application. Emit a helpful message, use a different default... I'd prefer the message.

kren1 replied:
I don't think users need to worry about `CommitEvery` unless they really want to. That is they want each instruction logged and committed because KLEE might crash. Emitting and error message is just whitenoise people ignore.

251 replied:
But you're mangling up responsibilities. If one sets `WriteInterval` to 100ms, you surely want to increase `CommitEvery`. We should keep it separated and simple.

251 replied:
Then remove a 0 from my example.

251 commented on scripts/build/klee.sh:
```diff
@@ -116,6 +116,17 @@ else
 fi
 
+if [[ "$TRAVIS_OS_NAME" = "linux" ]] ; then
+python3 -m pip install --user tabulate
```
indentation

251 commented on lib/Core/StatsTracker.cpp:
```diff
@@ -176,8 +181,10 @@ StatsTracker::StatsTracker(Executor &_executor, std::string _objectFilename,
     objectFilename(_objectFilename),
-    statsFile(0),
     istatsFile(0),
+    statsFile(0),
```
initialise in header file, not in constructor (unique_ptr?)

kren1 replied:
I tried using unique_ptr and I think it's very ugly as it needs to operate with sqlite3 function. For example you allocate most of the managed objects by passing `sqlite3**` to the allocation object.

251 replied:
I checked my version and it's also using raw pointers... Makes sense.

251 commented on lib/Core/StatsTracker.cpp:
```diff
@@ -267,7 +285,10 @@ StatsTracker::StatsTracker(Executor &_executor, std::string _objectFilename,
+  char *zErrMsg;
+  sqlite3_exec(statsFile, "END TRANSACTION", NULL, NULL, &zErrMsg);
+  sqlite3_finalize(insertStmt);
+  sqlite3_close(statsFile);
```
sqlite3_free is missing for zErrMsg

251 commented on lib/Core/StatsTracker.cpp:
```diff
@@ -421,63 +442,130 @@ void StatsTracker::markBranchVisited(ExecutionState *visitedTrue,
+        klee_error("Cannot create prepared statement! %s", sqlite3_errmsg(statsFile));
+        return;
+    }
+    sqlite3_exec(statsFile, "PRAGMA synchronous = OFF", NULL, NULL, &zErrMsg);
```
This might leak (check docs). I'm not sure how sqlite handles allocated zErrMsges. And again final free is missing.

251 commented on lib/Core/StatsTracker.cpp:
```diff
@@ -242,13 +254,19 @@ StatsTracker::StatsTracker(Executor &_executor, std::string _objectFilename,
+       std::stringstream error_s;
+       error_s << "Can't open database: " << sqlite3_errmsg(statsFile);
+       sqlite3_close(statsFile);
+       klee_error("%s", error_s.str().c_str());
```
Why don't you just put the sqlite3_errmsg here for s instead of the stringstream?

251 replied:
Wait... otherwise you would use it after _close. That shouldn't work. Keep it that way.

kren1 replied:
You mean, I get the same leak as below? This API is annoying ...

kren1 replied:
I reverted this.

251 commented on lib/Core/StatsTracker.cpp:
```diff
@@ -421,63 +442,130 @@ void StatsTracker::markBranchVisited(ExecutionState *visitedTrue,
+             if(writeCount % commitEvery == 0 ) {
+               char *zErrMsg = 0;
+               sqlite3_exec(statsFile, "END TRANSACTION", NULL, NULL, &zErrMsg);
+               sqlite3_exec(statsFile, "BEGIN TRANSACTION", NULL, NULL, &zErrMsg);
```
free? + error handling for step/reset is missing

MartinNowack replied:
Use `prepare/step/reset` for `END_TRANSACTION`.

MartinNowack replied:
And address @251's comment

251 commented on lib/Core/StatsTracker.cpp:
```diff
@@ -421,63 +437,131 @@ void StatsTracker::markBranchVisited(ExecutionState *visitedTrue,
+       char *zErrMsg = 0;
+       if(sqlite3_exec(statsFile, create.str().c_str(), NULL, NULL, &zErrMsg)) {         
+         klee_error("ERROR creating table: %s", zErrMsg);
+         sqlit3_free(zErrMsg);
```
Typo + klee_error exits before free...

kren1 replied:
So what should I do? Copy to a stack array and then free? That sounds weird, I wonder if there is an efficient way of doing it.

kren1 replied:
I refactored that into a function, so I think it should be fine now.

ccadar commented:
@kren1 I'll let @251 iterate and approve this request first and then I'll take a look

kren1 commented on lib/Core/StatsTracker.cpp:
```diff
@@ -395,6 +413,12 @@ void StatsTracker::framePopped(ExecutionState &es) {
 }
 
+std::string sqlite3ErrToStringAndFree(std::string prefix, char* sqlite3ErrMsg) {
+       std::stringstream stream_s;
```
s is string, I don't like ss for historical reasons, the other two are too similar to type names.

251 replied:
> s is string [...] too similar to type names

But you added the type to the name without even using `auto` :smile:  I'd drop it.

kren1 commented on lib/Core/StatsTracker.cpp:
```diff
@@ -421,63 +445,133 @@ void StatsTracker::markBranchVisited(ExecutionState *visitedTrue,
+              << "? "
+              << ")";
+    if(sqlite3_prepare_v2(statsFile, insert.str().c_str(), -1, &insertStmt, NULL)) {
+        klee_error("Cannot create prepared statement! %s", sqlite3_errmsg(statsFile));
```
`sqlite3_errmsg` is managed by sqlite3 so no need to free. Close DB should be handled by deconstructor. I don't care that early call to `exit` ignores those.

251 replied:
It is not managed because we kill it with exit... We could use the same cleanup method everywhere - that's what I meant.

kren1 replied:
But I can't do `sqlite3_free(sqlite3_errmsg(statsFile))` that screams double free.

251 replied:
No you shouldn't do that. Just close the DB properly here.

kren1 commented on lib/Core/StatsTracker.cpp:
```diff
@@ -421,63 +445,133 @@ void StatsTracker::markBranchVisited(ExecutionState *visitedTrue,
+             if(errCode != SQLITE_OK) klee_error("Error %d in sqlite3_reset function", errCode);
+
+             if(writeCount % commitEvery == 0 ) {
+               sqlite3_exec(statsFile, "END TRANSACTION", NULL, NULL, NULL);
```
Why add more code and make it slower for unrealistic overflow?

251 replied:
Do you think `writeCount == commitEvery` is slower than `writeCount % commitEvery == 0` or the one the single mov resulting from writeCount = 0 executed every commitEvery?!

kren1 replied:
Ahh I see, but then it's not really `writeCount`. I guess your approach is better.

kren1 commented on lib/Core/StatsTracker.cpp:
```diff
@@ -421,63 +445,133 @@ void StatsTracker::markBranchVisited(ExecutionState *visitedTrue,
-             << ")\n";
-  statsFile->flush();
+             int errCode = sqlite3_step(insertStmt);
+             if(errCode != SQLITE_DONE) klee_error("Error %d in sqlite3_step function", errCode);
```
Wait what do I need to free here to warant `sqlite3ErrToStringAndFree`? I don't see any sqlite3 allocated memory.

251 replied:
ah errCode... yeah I just copied it...

kren1 commented on tools/klee-stats/klee-stats:
```diff
@@ -65,49 +66,29 @@ def getLogFile(path):
-    return lo
+      #  self.lines = lines[1:]
+      self.conn = sqlite3.connect(fileName);
+      self.c = self.conn.cursor()
```
This is not really easy and here is not the right place to do it.

kren1 commented:
I think I addressed all your comments. Also pushed a `docker run` wrapper script. I don't know if we want it here, but I find it convenient.

251 commented:
@ccadar @MartinNowack After endless iterations I think we reached a fixed point (I don't want to read it anymore and @kren1 doesn't want to fix it anymore :smile:). Looks good to me now.

MartinNowack commented on lib/Core/StatsTracker.h:
```diff
@@ -37,7 +38,11 @@ namespace klee {
 
-    std::unique_ptr<llvm::raw_fd_ostream> statsFile, istatsFile;
+    std::unique_ptr<llvm::raw_fd_ostream> istatsFile;
+    sqlite3 *statsFile;
```
Needs to be `nullptr` initialised

MartinNowack commented on lib/Core/StatsTracker.h:
```diff
@@ -37,7 +38,11 @@ namespace klee {
-    std::unique_ptr<llvm::raw_fd_ostream> statsFile, istatsFile;
+    std::unique_ptr<llvm::raw_fd_ostream> istatsFile;
+    sqlite3 *statsFile;
+    sqlite3_stmt *insertStmt;
```
Needs to be `nullptr` initialised

MartinNowack commented on lib/Core/StatsTracker.h:
```diff
@@ -37,7 +38,11 @@ namespace klee {
+    std::unique_ptr<llvm::raw_fd_ostream> istatsFile;
+    sqlite3 *statsFile;
+    sqlite3_stmt *insertStmt;
+    unsigned writeCount;
```
Rename to be clear about being specific to `statsFile`

MartinNowack commented on lib/Core/StatsTracker.h:
```diff
@@ -37,7 +38,11 @@ namespace klee {
+    sqlite3 *statsFile;
+    sqlite3_stmt *insertStmt;
+    unsigned writeCount;
+    unsigned commitEvery;
```
Rename to be clear about being specific to `statsFile`

MartinNowack commented on lib/Core/StatsTracker.cpp:
```diff
@@ -420,63 +443,130 @@ void StatsTracker::markBranchVisited(ExecutionState *visitedTrue,
+             sqlite3_reset(insertStmt);
+             if(writeCount % commitEvery == 0 ) {
+               char *zErrMsg = 0;
+               sqlite3_exec(statsFile, "END TRANSACTION", NULL, NULL, &zErrMsg);
```
Use `prepare/step/reset` for `END_TRANSACTION`.

kren1 replied:
why?

MartinNowack commented on lib/Core/StatsTracker.cpp:
```diff
@@ -420,63 +443,130 @@ void StatsTracker::markBranchVisited(ExecutionState *visitedTrue,
+             sqlite3_bind_int64(insertStmt, 11, elapsed());
+             sqlite3_bind_int64(insertStmt, 12, stats::coveredInstructions);
+             sqlite3_bind_int64(insertStmt, 13, stats::uncoveredInstructions);
+             sqlite3_bind_int64(insertStmt, 14, stats::queryTime / 1000000.);
```
`sqlite3_bind_double()`

kren1 replied:
@251 and I decided to keep this as an int and do the division in `klee-stats`, it makes more sense that way. This is old code.

MartinNowack commented on lib/Core/StatsTracker.cpp:
```diff
@@ -420,63 +443,130 @@ void StatsTracker::markBranchVisited(ExecutionState *visitedTrue,
+             sqlite3_bind_int64(insertStmt, 12, stats::coveredInstructions);
+             sqlite3_bind_int64(insertStmt, 13, stats::uncoveredInstructions);
+             sqlite3_bind_int64(insertStmt, 14, stats::queryTime / 1000000.);
+             sqlite3_bind_int64(insertStmt, 15, stats::solverTime / 1000000.);
```
`sqlite3_bind_double()`

MartinNowack commented on lib/Core/StatsTracker.cpp:
```diff
@@ -420,63 +443,130 @@ void StatsTracker::markBranchVisited(ExecutionState *visitedTrue,
+             sqlite3_bind_int64(insertStmt, 13, stats::uncoveredInstructions);
+             sqlite3_bind_int64(insertStmt, 14, stats::queryTime / 1000000.);
+             sqlite3_bind_int64(insertStmt, 15, stats::solverTime / 1000000.);
+             sqlite3_bind_int64(insertStmt, 16, stats::cexCacheTime / 1000000.);
```
`sqlite3_bind_double()`

MartinNowack commented on lib/Core/StatsTracker.cpp:
```diff
@@ -420,63 +443,130 @@ void StatsTracker::markBranchVisited(ExecutionState *visitedTrue,
+             sqlite3_bind_int64(insertStmt, 14, stats::queryTime / 1000000.);
+             sqlite3_bind_int64(insertStmt, 15, stats::solverTime / 1000000.);
+             sqlite3_bind_int64(insertStmt, 16, stats::cexCacheTime / 1000000.);
+             sqlite3_bind_int64(insertStmt, 17, stats::forkTime / 1000000.);
```
`sqlite3_bind_double()`

MartinNowack commented on lib/Core/StatsTracker.cpp:
```diff
@@ -420,63 +443,130 @@ void StatsTracker::markBranchVisited(ExecutionState *visitedTrue,
+             sqlite3_bind_int64(insertStmt, 15, stats::solverTime / 1000000.);
+             sqlite3_bind_int64(insertStmt, 16, stats::cexCacheTime / 1000000.);
+             sqlite3_bind_int64(insertStmt, 17, stats::forkTime / 1000000.);
+             sqlite3_bind_int64(insertStmt, 18, stats::resolveTime / 1000000.);
```
`sqlite3_bind_double()`

MartinNowack commented on lib/Core/StatsTracker.cpp:
```diff
@@ -420,63 +443,130 @@ void StatsTracker::markBranchVisited(ExecutionState *visitedTrue,
+             sqlite3_bind_int64(insertStmt, 20, stats::queryCexCacheHits);
 #ifdef KLEE_ARRAY_DEBUG
-             << "," << stats::arrayHashTime / 1000000.
+             sqlite3_bind_int64(insertStmt, 21, stats::arrayHashTime / 1000000.);
```
`sqlite3_bind_double()`

kren1 replied:
I'll keep this consistent with the others.

MartinNowack commented on lib/Core/StatsTracker.cpp:
```diff
@@ -420,63 +443,130 @@ void StatsTracker::markBranchVisited(ExecutionState *visitedTrue,
+             << "WallTime int,"
+             << "CoveredInstructions int,"
+             << "UncoveredInstructions int,"
+             << "QueryTime int,"
```
`DOUBLE`

MartinNowack commented on lib/Core/StatsTracker.cpp:
```diff
@@ -420,63 +443,130 @@ void StatsTracker::markBranchVisited(ExecutionState *visitedTrue,
+             << "CoveredInstructions int,"
+             << "UncoveredInstructions int,"
+             << "QueryTime int,"
+             << "SolverTime int,"
```
`DOUBLE`

MartinNowack commented on lib/Core/StatsTracker.cpp:
```diff
@@ -420,63 +443,130 @@ void StatsTracker::markBranchVisited(ExecutionState *visitedTrue,
+             << "UncoveredInstructions int,"
+             << "QueryTime int,"
+             << "SolverTime int,"
+             << "CexCacheTime int,"
```
`DOUBLE`

MartinNowack commented on lib/Core/StatsTracker.cpp:
```diff
@@ -420,63 +443,130 @@ void StatsTracker::markBranchVisited(ExecutionState *visitedTrue,
+             << "QueryTime int,"
+             << "SolverTime int,"
+             << "CexCacheTime int,"
+             << "ForkTime int,"
```
`DOUBLE`

MartinNowack commented on lib/Core/StatsTracker.cpp:
```diff
@@ -420,63 +443,130 @@ void StatsTracker::markBranchVisited(ExecutionState *visitedTrue,
+             << "SolverTime int,"
+             << "CexCacheTime int,"
+             << "ForkTime int,"
+             << "ResolveTime int,"
```
`DOUBLE`

MartinNowack commented on lib/Core/StatsTracker.cpp:
```diff
@@ -420,63 +443,130 @@ void StatsTracker::markBranchVisited(ExecutionState *visitedTrue,
+             << "ResolveTime int,"
+             << "QueryCexCacheMisses int,"
+#ifdef KLEE_ARRAY_DEBUG
+	     << "ArrayHashTime int,"
```
`REAL`

MartinNowack commented on lib/Core/StatsTracker.cpp:
```diff
@@ -420,63 +443,130 @@ void StatsTracker::markBranchVisited(ExecutionState *visitedTrue,
+              << "ForkTime ,"
+              << "ResolveTime ,"
+              << "QueryCexCacheMisses ,"
+#ifdef KLEE_ARRAY_DEBUG
```
`#ifdef` including content should be after `QueryCexCacheHits`

kren1 replied:
Btw this is how it was before. Also if `ifdef` is last it makes commans annoying.

kren1 replied:
Yes, I really don't want to introduce an `elif`

MartinNowack replied:
No it wasn't done before like that, the order was
`QueryCexCacheMisses, QueryCexCacheHits, ArrayHashTime.`

kren1 replied:
Ok you are right, my bad. Still the order it is now, is the cleanest way to write this query.  Since sqlite is a relational database the actual order of columns doesn't matter.

MartinNowack commented on lib/Core/StatsTracker.cpp:
```diff
@@ -420,63 +443,130 @@ void StatsTracker::markBranchVisited(ExecutionState *visitedTrue,
+             << "ForkTime int,"
+             << "ResolveTime int,"
+             << "QueryCexCacheMisses int,"
+#ifdef KLEE_ARRAY_DEBUG
```
Should be after `QueryCexCacheHits`

MartinNowack commented on lib/Core/StatsTracker.cpp:
```diff
@@ -421,63 +445,133 @@ void StatsTracker::markBranchVisited(ExecutionState *visitedTrue,
-             << "'QueryCexCacheHits',"
+  std::stringstream create,  insert;
+  create << "CREATE TABLE stats ";
+  create     << "(Instructions int,"
```
please replace `int` with `INTEGER` where appropriate

MartinNowack commented:
@kren1 Github is a behaving strangely - many comments I made are marked as outdated but they aren't.

kren1 commented:
@MartinNowack yes github is behaving strangly, but you were looking at old version of the code (happened to @251 too). For example things are nullptr initialized and I addressed @251's comments. Some of your comments are still valid, will address those.

Also what's wrong with `sqlite3_exec`? I find it much cleaner for some things.

kren1 commented:
@MartinNowack I addressed the comments that I like, rebased and fixed weird git issue where there were some weird whitespace diffs. Probably #996 didn't do wonders for this patch.

jbuening commented:
@kren1 I am sorry that #996 made it harder for you to rebase. Or was there another issue that you meant to hint at?

I did not take a closer look at this set of changes, but wouldn't it be a good time to get rid of `numObjects` with this PR instead of hardcoding it to 0?

kren1 commented:
@jbuening don't worry about it. There was just an issue where the indentation changed for some reason in sizeable amount of StatsTracker.cpp, I didn't investigate further.

MartinNowack commented on lib/Core/StatsTracker.cpp:
```diff
@@ -240,15 +250,18 @@ StatsTracker::StatsTracker(Executor &_executor, std::string _objectFilename,
   if (OutputStats) {
-    statsFile = executor.interpreterHandler->openOutputFile("run.stats");
-    if (statsFile) {
+    auto db_filename = executor.interpreterHandler->getOutputFilename("run.stats");
```
Wooow an `auto` :car:

MartinNowack commented:
@kren1 The "problem" with `sqlite3_exec` is that it internally executes `sqlite3_prepare_v2()`,  `sqlite3_step()`, and `sqlite3_finalize()` https://sqlite.org/c3ref/exec.html. That is ok (and nicer!!) for non-performance-critical code but not in this case. For a small `commit-after-n-instructions` it's going to be executed too often, therefore it would be nice to move the `sqlite3_prepare_v2()` outside of the critical loop.

kren1 commented:
@MartinNowack  If you are commiting every instructions you are going to take a performance hit anyway, a small speedup of putting the `sqlite3_prepare_v2()` outside the loop should be insignificant. For the wast majority of cases this isn't performance critical code.

MartinNowack commented:
@kren1 Still, you don't need to make it even slower.

kren1 commented:
@MartinNowack  I do, because it makes the rest of the code messy. If you are this worried about this, I would prefer to prevent people from committing at every instruction.

MartinNowack commented:
@kren1 People might have reasons to commit with a higher ratio. By the way, @251 suggested to merge the two queries into one. If this PR goes in, it is going to replace a vital part of KLEE, so yes I want to have performance there.

251 commented:
> People might have reasons to commit with a higher ratio.

Debugging is pretty annoying with unflushed buffers (or unwritten stats). I think this is a valid use case and I agree with @MartinNowack.
